src/server/string_family.cc

```diff
@@ -413,67 +409,86 @@ OpResult<array<int64_t, 5>> OpThrottle(const OpArgs& op_args, const string_view
     return OpStatus::INVALID_INT;
   }
-  const int64_t allow_at_ms = new_tat_ms - delay_variation_tolerance_ms;
 
-  if (allow_at_ms >= 0 ? now_ms < INT64_MIN + allow_at_ms : now_ms > INT64_MAX + allow_at_ms) {
+  // The cutoff point before which a request is rejected (throttled) and at or after which a request
+  // is accepted.
+  const int64_t allow_at_ns = new_tat_ns - delay_variation_tolerance_ns;
+
+  if (allow_at_ns >= 0 ? now_ns < INT64_MIN + allow_at_ns : now_ns > INT64_MAX + allow_at_ns) {
     return OpStatus::INVALID_INT;
   }
-  const int64_t diff_ms = now_ms - allow_at_ms;
+  const int64_t diff_ns = now_ns - allow_at_ns;
 
-  const bool limited = diff_ms < 0;
-  int64_t ttl_ms;
+  const bool limited = diff_ns < 0;
+  int64_t ttl_ns;
   if (limited) {
-    if (increment_ms <= delay_variation_tolerance_ms) {
-      if (diff_ms == INT64_MIN) {
+    if (increment_ns <= delay_variation_tolerance_ns) {
+      if (diff_ns == INT64_MIN) {
         return OpStatus::INVALID_INT;
       }
-      retry_after_ms = -diff_ms;
+      retry_after_ms = -(diff_ns / 1000000);
+      if (-diff_ns > 0) {
+        retry_after_ms += 1;
+      }
     }
 
-    if (now_ms >= 0 ? tat_ms < INT64_MIN + now_ms : tat_ms > INT64_MAX + now_ms) {
+    if (now_ns >= 0 ? tat_ns < INT64_MIN + now_ns : tat_ns > INT64_MAX + now_ns) {
       return OpStatus::INVALID_INT;
     }
-    ttl_ms = tat_ms - now_ms;
+    ttl_ns = tat_ns - now_ns;
   } else {
-    if (now_ms >= 0 ? new_tat_ms < INT64_MIN + now_ms : new_tat_ms > INT64_MAX + now_ms) {
+    if (now_ns >= 0 ? new_tat_ns < INT64_MIN + now_ns : new_tat_ns > INT64_MAX + now_ns) {
       return OpStatus::INVALID_INT;
     }
-    ttl_ms = new_tat_ms - now_ms;
+    ttl_ns = new_tat_ns - now_ns;
   }
 
-  if (ttl_ms < delay_variation_tolerance_ms - INT64_MAX) {
+  if (ttl_ns < delay_variation_tolerance_ns - INT64_MAX) {
     return OpStatus::INVALID_INT;
   }
-  const int64_t next_ms = delay_variation_tolerance_ms - ttl_ms;
-  if (next_ms > -emission_interval_ms) {
-    remaining = next_ms / emission_interval_ms;
+  const int64_t next_ns = delay_variation_tolerance_ns - ttl_ns;
+  if (next_ns > -emission_interval_ns) {
+    remaining = next_ns / emission_interval_ns;
+  }
+  reset_after_ms = ttl_ns / 1000000;
+  if (ttl_ns) {
+    reset_after_ms += 1;
   }
-  reset_after_ms = ttl_ms;
 
   if (!limited) {
+    // Although most computation so far is in nanoseconds, we must store expiry as milliseconds.
+    // While this causes loss of precision, the value stored against the throttle key is still in
+    // the nanosecond units. When the key is loaded, that value will be read and used as tat_ns. The
+    // loss of precision will cause the throttle key to be expired a bit earlier than expected, so
+    // to make up, we extend its expiry by 1 millisecond. Extending the key life does not break
+    // behavior because the tat_ns value will be used to check for throttling.
+    int64_t new_tat_ms = new_tat_ns / 1000000;
```

Comment:
this is not how you round up division by N. this is how you do it:
`x= (y + N -1) / N`